<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Introduction to GenotypeMixtures • GenotypeMixtures</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Introduction to GenotypeMixtures" />
<meta property="og:description" content="GenotypeMixtures" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GenotypeMixtures</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction.html">Introduction to GenotypeMixtures</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<script src="introduction_files/header-attrs-2.7/header-attrs.js"></script>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to GenotypeMixtures</h1>
            
            <h4 class="date">2021-06-21</h4>
      
      
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<p>GenotypeMixtures is a handy package that builds on the souporcell package (Heaton et al. 2020), to stitch together genotypes across multiple single cell genomics experiments with an overlapping mixture experimental design…</p>
<p>Install GenotypeMixtures from github. Requires devtools.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#devtools::install_github(&quot;bjstewart1/GenotypeMixtures&quot;)</span></span></code></pre></div>
<p>Load GenotypeMixtures</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenotypeMixtures)</span></code></pre></div>
<p>Experimental designs can be read in using this function if you point at a .csv Alternatively you can read in the .csv however you like, or construct from another file The experiments (10X channels should be rows, and the donors should be columns. Membership is denoted by 1 vs 0.</p>
<pre><code>#&gt; reading experimental design
#&gt; Using channel as id variables</code></pre>
<p><img src="/lustre/scratch117/cellgen/team297/bs16/GenotypeMixtures/docs/articles/introduction_files/figure-html/unnamed-chunk-4-1.png" width="700" /></p>
<p>We can also read in the locations of the souporcell directories</p>
<pre><code>#&gt; reading souporcell file locations
#&gt;             channel
#&gt; 1   CG_SB_NB8768284
#&gt; 2 WSSS_A_KID9369817
#&gt; 3 WSSS_A_KID9369818
#&gt; 4 WSSS_A_KID9369819
#&gt; 5 WSSS_A_KID9369820
#&gt; 6 WSSS_A_KID9369821
#&gt;                                                                                                                       SOC_location
#&gt; 1     /lustre/scratch117/cellgen/team297/bs16/10X_data/kidney_glom/10x3p_cells/CG_SB_NB8768284/CG_SB_NB8768284_SOC_common_variants
#&gt; 2 /lustre/scratch117/cellgen/team297/bs16/10X_data/kidney_glom/10x3p_cells/WSSS_A_KID9369817/WSSS_A_KID9369817_SOC_common_variants
#&gt; 3 /lustre/scratch117/cellgen/team297/bs16/10X_data/kidney_glom/10x3p_cells/WSSS_A_KID9369818/WSSS_A_KID9369818_SOC_common_variants
#&gt; 4 /lustre/scratch117/cellgen/team297/bs16/10X_data/kidney_glom/10x3p_cells/WSSS_A_KID9369819/WSSS_A_KID9369819_SOC_common_variants
#&gt; 5 /lustre/scratch117/cellgen/team297/bs16/10X_data/kidney_glom/10x3p_cells/WSSS_A_KID9369820/WSSS_A_KID9369820_SOC_common_variants
#&gt; 6 /lustre/scratch117/cellgen/team297/bs16/10X_data/kidney_glom/10x3p_cells/WSSS_A_KID9369821/WSSS_A_KID9369821_SOC_common_variants</code></pre>
<p>Now we plug this into the main function which constructs a genotype cluster graph</p>
<pre><code>#&gt; checking files
#&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |=                                                                     |   2%
#&gt; 
#&gt;    *****       ***   vcfR   ***       *****
#&gt;    This is vcfR 1.12.0 
#&gt;      browseVignettes(&#39;vcfR&#39;) # Documentation
#&gt;      citation(&#39;vcfR&#39;) # Citation
#&gt;    *****       *****      *****       *****
#&gt; 
  |                                                                            
  |==                                                                    |   3%
  |                                                                            
  |===                                                                   |   5%
  |                                                                            
  |====                                                                  |   6%
  |                                                                            
  |=====                                                                 |   8%
  |                                                                            
  |=======                                                               |   9%
  |                                                                            
  |========                                                              |  11%
  |                                                                            
  |=========                                                             |  12%
  |                                                                            
  |==========                                                            |  14%
  |                                                                            
  |===========                                                           |  16%
  |                                                                            
  |============                                                          |  17%
  |                                                                            
  |=============                                                         |  19%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |===============                                                       |  22%
  |                                                                            
  |================                                                      |  23%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |===================                                                   |  27%
  |                                                                            
  |====================                                                  |  28%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |======================                                                |  31%
  |                                                                            
  |=======================                                               |  33%
  |                                                                            
  |========================                                              |  34%
  |                                                                            
  |=========================                                             |  36%
  |                                                                            
  |==========================                                            |  38%
  |                                                                            
  |===========================                                           |  39%
  |                                                                            
  |============================                                          |  41%
  |                                                                            
  |==============================                                        |  42%
  |                                                                            
  |===============================                                       |  44%
  |                                                                            
  |================================                                      |  45%
  |                                                                            
  |=================================                                     |  47%
  |                                                                            
  |==================================                                    |  48%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |====================================                                  |  52%
  |                                                                            
  |=====================================                                 |  53%
  |                                                                            
  |======================================                                |  55%
  |                                                                            
  |=======================================                               |  56%
  |                                                                            
  |========================================                              |  58%
  |                                                                            
  |==========================================                            |  59%
  |                                                                            
  |===========================================                           |  61%
  |                                                                            
  |============================================                          |  62%
  |                                                                            
  |=============================================                         |  64%
  |                                                                            
  |==============================================                        |  66%
  |                                                                            
  |===============================================                       |  67%
  |                                                                            
  |================================================                      |  69%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |==================================================                    |  72%
  |                                                                            
  |===================================================                   |  73%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |======================================================                |  77%
  |                                                                            
  |=======================================================               |  78%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |=========================================================             |  81%
  |                                                                            
  |==========================================================            |  83%
  |                                                                            
  |===========================================================           |  84%
  |                                                                            
  |============================================================          |  86%
  |                                                                            
  |=============================================================         |  88%
  |                                                                            
  |==============================================================        |  89%
  |                                                                            
  |===============================================================       |  91%
  |                                                                            
  |=================================================================     |  92%
  |                                                                            
  |==================================================================    |  94%
  |                                                                            
  |===================================================================   |  95%
  |                                                                            
  |====================================================================  |  97%
  |                                                                            
  |===================================================================== |  98%
  |                                                                            
  |======================================================================| 100%
#&gt; 
#&gt; Attaching package: &#39;igraph&#39;
#&gt; The following objects are masked from &#39;package:stats&#39;:
#&gt; 
#&gt;     decompose, spectrum
#&gt; The following object is masked from &#39;package:base&#39;:
#&gt; 
#&gt;     union
#&gt; the membership graph is a collection of complete subgraphs as expected</code></pre>
<p>Now we can plot the graph which stitches together the genotypes</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>genotype_clustering_output<span class="sc">$</span>graph_plot</span></code></pre></div>
<p><img src="/lustre/scratch117/cellgen/team297/bs16/GenotypeMixtures/docs/articles/introduction_files/figure-html/unnamed-chunk-7-1.png" width="700" /></p>
<p>Now we can plot the membership matrix which tells us which of our genotypes belongs to which mixtures</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>genotype_clustering_output<span class="sc">$</span>membership_plot</span></code></pre></div>
<p><img src="/lustre/scratch117/cellgen/team297/bs16/GenotypeMixtures/docs/articles/introduction_files/figure-html/unnamed-chunk-8-1.png" width="700" /></p>
<p>Now we can plot genotype VAFs - this is a useful diagnostic plot; matching genotypes should have their variants along the diagnonal.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_cross_vaf</span>(<span class="at">experiment_1_path =</span> file_locations[<span class="dv">2</span>, <span class="dv">2</span>], </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">experiment_2_path =</span> file_locations[<span class="dv">3</span>,<span class="dv">2</span>], </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">experiment_1_name =</span> file_locations[<span class="dv">2</span>,<span class="dv">1</span>],</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">experiment_2_name =</span> file_locations[<span class="dv">3</span>,<span class="dv">1</span>])</span></code></pre></div>
<p><img src="/lustre/scratch117/cellgen/team297/bs16/GenotypeMixtures/docs/articles/introduction_files/figure-html/unnamed-chunk-9-1.png" width="700" /></p>
<p>We can map these computed genotypes back to the original genotypes in our experimental design</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>cluster_mapping <span class="ot">&lt;-</span> <span class="fu">membership_map</span>(<span class="at">experimental_design =</span> experimental_design,</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">graph_output =</span>  genotype_clustering_output)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(cluster_mapping)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;               channel          SOC_cluster genotype_cluster genotype</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 64 WSSS_A_KID10063749 WSSS_A_KID10063749_3               15     K272</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 65 WSSS_A_KID10063750 WSSS_A_KID10063750_5               15     K272</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 66 WSSS_A_KID10063745 WSSS_A_KID10063745_1                7     K273</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 67 WSSS_A_KID10063746 WSSS_A_KID10063746_2                7     K273</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 68 WSSS_A_KID10063749 WSSS_A_KID10063749_0                7     K273</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 69 WSSS_A_KID10063750 WSSS_A_KID10063750_1                7     K273</span></span></code></pre></div>
<p>Finally we can assign single cells across our experiments to genotype - feed the output of membership_map() to cells_to_genotypes The output of this can be easily added to the metadata of your single cell experiment/seurat/anndata object</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>cell_assignments <span class="ot">&lt;-</span> <span class="fu">cells_to_genotypes</span>(<span class="at">SOC_locations =</span> file_locations, </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">membership_mat =</span>cluster_mapping)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(cell_assignments)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                   barcode  status assignment            channel genotype</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 166160 TTTGTTGAGACAGCTG-1 singlet          0 WSSS_A_KID10063750     K269</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 166161 TTTGTTGAGTTTCTTC-1 singlet          5 WSSS_A_KID10063750     K272</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 166162 TTTGTTGCACCTGTCT-1 singlet          5 WSSS_A_KID10063750     K272</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 166163 TTTGTTGCACGTCTCT-1 singlet          0 WSSS_A_KID10063750     K269</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 166164 TTTGTTGGTCACGACC-1 singlet          4 WSSS_A_KID10063750     K270</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 166165 TTTGTTGTCGCCGAAC-1 singlet          4 WSSS_A_KID10063750     K270</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                 SOC_cluster</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 166160 WSSS_A_KID10063750_0</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 166161 WSSS_A_KID10063750_5</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 166162 WSSS_A_KID10063750_5</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 166163 WSSS_A_KID10063750_0</span></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 166164 WSSS_A_KID10063750_4</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 166165 WSSS_A_KID10063750_4</span></span></code></pre></div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer>
      <div class="copyright">
  <p>Developed by Benjamin Stewart.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>
