<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to GenotypeMixtures • GenotypeMixtures</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to GenotypeMixtures">
<meta property="og:description" content="GenotypeMixtures">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GenotypeMixtures</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Introduction to GenotypeMixtures</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="introduction_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to GenotypeMixtures</h1>
            
            <h4 class="date">2021-11-04</h4>
      
      
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<p>GenotypeMixtures is a handy package that builds on the souporcell package (Heaton et al. 2020), to stitch together genotypes across multiple single cell genomics experiments with an overlapping mixture experimental design…</p>
<p>Install GenotypeMixtures from github. Requires devtools.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#devtools::install_github("bjstewart1/GenotypeMixtures")</span></code></pre></div>
<p>Load GenotypeMixtures</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GenotypeMixtures</span><span class="op">)</span></code></pre></div>
<p>Experimental designs can be read in using this function if you point at a .csv Alternatively you can read in the .csv however you like, or construct from another file The experiments (10X channels (mixtures) should be rows, and the donors/genotypes should be columns. Membership is denoted by 1 vs 0.</p>
<pre><code>#&gt; reading experimental design
#&gt; Using channel as id variables</code></pre>
<p><img src="introduction_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>We can also read in the locations of the souporcell directories The first column should be the mixture name, the second column should be the path to the soup or cell directory There is some built in dummy vcf files in the package for this vignette</p>
<pre><code>#&gt;      channel
#&gt; 1 mixtures_1
#&gt; 2 mixtures_2
#&gt; 3 mixtures_3
#&gt; 4 mixtures_4
#&gt; 5 mixtures_5
#&gt;                                                          SOC_directory
#&gt; 1 /Users/runner/work/_temp/Library/GenotypeMixtures/extdata/mixtures_1
#&gt; 2 /Users/runner/work/_temp/Library/GenotypeMixtures/extdata/mixtures_2
#&gt; 3 /Users/runner/work/_temp/Library/GenotypeMixtures/extdata/mixtures_3
#&gt; 4 /Users/runner/work/_temp/Library/GenotypeMixtures/extdata/mixtures_4
#&gt; 5 /Users/runner/work/_temp/Library/GenotypeMixtures/extdata/mixtures_5</code></pre>
<p>Now we plug this into the main function which constructs a genotype cluster graph</p>
<pre><code>#&gt; checking files
#&gt; 
  |                                                                            
  |                                                                      |   0%
#&gt; 
#&gt;    *****       ***   vcfR   ***       *****
#&gt;    This is vcfR 1.12.0 
#&gt;      browseVignettes('vcfR') # Documentation
#&gt;      citation('vcfR') # Citation
#&gt;    *****       *****      *****       *****
#&gt; 
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |======================================================================| 100%
#&gt; the membership graph is a collection of complete subgraphs as expected</code></pre>
<p>Now we can plot the graph which stitches together the genotypes</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genotype_clustering_output</span><span class="op">$</span><span class="va">graph_plot</span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Now we can plot the membership matrix which tells us which of our genotypes belongs to which mixtures</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genotype_clustering_output</span><span class="op">$</span><span class="va">membership_plot</span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Now we can plot genotype VAFs - this is a useful diagnostic plot; matching genotypes should have their variants along the diagnonal. This is synthetic data, but real data should look reasonably similar to this</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_cross_vaf.html">plot_cross_vaf</a></span><span class="op">(</span>experiment_1_path <span class="op">=</span> <span class="va">file_locations</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">]</span>, 
               experiment_2_path <span class="op">=</span> <span class="va">file_locations</span><span class="op">[</span><span class="fl">3</span>,<span class="fl">2</span><span class="op">]</span>, 
               experiment_1_name <span class="op">=</span> <span class="va">file_locations</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">]</span>,
               experiment_2_name <span class="op">=</span> <span class="va">file_locations</span><span class="op">[</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>We can map these computed genotypes back to the original genotypes in our experimental design</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cluster_mapping</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/membership_map.html">membership_map</a></span><span class="op">(</span>experimental_design <span class="op">=</span> <span class="va">experimental_design</span>,
                                  graph_output <span class="op">=</span>  <span class="va">genotype_clustering_output</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">cluster_mapping</span><span class="op">)</span>
<span class="co">#&gt;       channel  SOC_cluster genotype_cluster   genotype</span>
<span class="co">#&gt; 26 mixtures_2 mixtures_2_4                3 genotype_8</span>
<span class="co">#&gt; 27 mixtures_4 mixtures_4_4                3 genotype_8</span>
<span class="co">#&gt; 28 mixtures_5 mixtures_5_4                3 genotype_8</span>
<span class="co">#&gt; 29 mixtures_3 mixtures_3_4                1 genotype_9</span>
<span class="co">#&gt; 30 mixtures_4 mixtures_4_5                1 genotype_9</span>
<span class="co">#&gt; 31 mixtures_5 mixtures_5_5                1 genotype_9</span></code></pre></div>
<p>Finally we can assign single cells across our experiments to genotype - feed the output of membership_map() to cells_to_genotypes The output of this can be easily added to the metadata of your single cell experiment/seurat/anndata object</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cell_assignments</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cells_to_genotypes.html">cells_to_genotypes</a></span><span class="op">(</span>SOC_locations <span class="op">=</span> <span class="va">file_locations</span>, 
                                       membership_mat <span class="op">=</span><span class="va">cluster_mapping</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">cell_assignments</span><span class="op">)</span>
<span class="co">#&gt;                 barcodes  status assignment    channel   genotype  SOC_cluster</span>
<span class="co">#&gt; 23993 CAGCCGATCGTAGGAG-1 singlet          3 mixtures_5 genotype_7 mixtures_5_3</span>
<span class="co">#&gt; 23994 ACACTGAAGCTCCTTC-1 singlet          0 mixtures_5 genotype_2 mixtures_5_0</span>
<span class="co">#&gt; 23995 CATGCCTCACCTCGGA-1 singlet          5 mixtures_5 genotype_9 mixtures_5_5</span>
<span class="co">#&gt; 23996 CCTACCACAAATACAG-1 singlet          0 mixtures_5 genotype_2 mixtures_5_0</span>
<span class="co">#&gt; 23997 AGATTGCTCTCCTATA-1 singlet          0 mixtures_5 genotype_2 mixtures_5_0</span>
<span class="co">#&gt; 23998 CCAGCGACATGGTTGT-1 singlet          1 mixtures_5 genotype_4 mixtures_5_1</span></code></pre></div>
<p>The package can also output an experimental design with varying levels of density</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dense_design</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_overlapping_mixture.html">make_overlapping_mixture</a></span><span class="op">(</span>n_mixtures <span class="op">=</span> <span class="fl">12</span>, n_genotypes <span class="op">=</span> <span class="fl">7</span>, density <span class="op">=</span> <span class="fl">1</span> <span class="op">)</span>
<span class="va">medium_density_design</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_overlapping_mixture.html">make_overlapping_mixture</a></span><span class="op">(</span>n_mixtures <span class="op">=</span> <span class="fl">12</span>, n_genotypes <span class="op">=</span> <span class="fl">7</span>, density <span class="op">=</span> <span class="fl">0.5</span> <span class="op">)</span>
<span class="va">sparse_design</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_overlapping_mixture.html">make_overlapping_mixture</a></span><span class="op">(</span>n_mixtures <span class="op">=</span> <span class="fl">12</span>, n_genotypes <span class="op">=</span> <span class="fl">7</span>, density <span class="op">=</span> <span class="fl">0</span> <span class="op">)</span></code></pre></div>
<p>This is a dense design</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_experimental_design.html">plot_experimental_design</a></span><span class="op">(</span><span class="va">dense_design</span><span class="op">)</span>
<span class="co">#&gt; Using channel as id variables</span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>This is a medium design</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_experimental_design.html">plot_experimental_design</a></span><span class="op">(</span><span class="va">medium_density_design</span><span class="op">)</span>
<span class="co">#&gt; Using channel as id variables</span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>This is a sparse design</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_experimental_design.html">plot_experimental_design</a></span><span class="op">(</span><span class="va">sparse_design</span><span class="op">)</span>
<span class="co">#&gt; Using channel as id variables</span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Benjamin Stewart.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
